### DISCUSSION TITLE
Cahn-Hilliard and NeumannBC

### DISCUSSION THREAD

**Discussion by souzanha** [idaholab/moose#17345] [Q&A Modules: Phase field] [answered]
Hi all! 
I'm working with the Cahn-Hilliard kernels and want to include a BC with an outward flux of a species. 
At the moment, I'm trying with a constant flux using a NeumannBC. The files compile and the solution converges, but the concentration is not changing. Any suggestions to what could be wrong?

[Mesh]
  type = GeneratedMesh # Can generate simple lines, rectangles and rectangular prisms
  dim = 2 # Dimension of the mesh
  nx = 100 # Number of elements in the x direction
  ny = 100 # Number of elements in the y direction
  xmin = 0
  xmax = 100e-9
  ymin= 0
  ymax = 100e-9
[]

[Variables]
  [./c]
  [../]

  [./w]
  [../]
[]

[ICs]
  [./c]
    type = ConstantIC
    variable = c
    value = 1
  [../]
[]

#[BCs]
 #[./right]
    #type = ADButlerVolmerBC
    #variable = c
    #boundary = right
    #w = w
 #[../]
#[]

**[BCs]
 [./right]
   type = NeumannBC
   variable = w
   boundary = right
   value = -0.1
 [../]
[]**

[Materials]
  [./constants]
    type = GenericConstantMaterial
    prop_names = 'R T omega Vm D'
    prop_values = '8.3145 298 12e3 43.8e-6 1e-15'
  [../]

  [./free_energy]
    type = DerivativeParsedMaterial
    f_name = f
    args = 'c'
    constant_names = 'R T omega Vm'
    constant_expressions = '8.3145 298 12e3 43.8e-6'
    function = '((R*T/Vm)*(c*plog(c,1e-3)+(1-c)*plog(1-c,1e-3)))+((omega/Vm)*c*(1-c))'
    derivative_order = 2
  [../]

  [./mobility]
    type = DerivativeParsedMaterial
    f_name = mob
    function = '(c*(1-c)*D*Vm)/(R*T)'
    args = 'c'
    constant_names  = 'R T Vm D'
    constant_expressions = '8.3145 298 43.8e-6 1e-15'
    derivative_order = 1
  [../]
[]

**[Kernels]
  [./time]
    type=CoupledTimeDerivative
    variable = w
    v = c
  [../]

  [./wres]
    type=SplitCHWRes
    variable = w
    mob_name = mob
  [../]

  [./c_res]
    type=SplitCHParsed
    variable = c
    w = w
    f_name = f
    kappa_name = 1.68e-10
  [../]
[]**

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = bdf2

  solve_type = NEWTON

  l_max_its = 30
  nl_max_its = 10
  l_tol = 1.0e-4
  nl_rel_tol = 1.0e-6
  nl_abs_tol = 1.0e-6

  **dt = 1e-3
  end_time = 10**
[]

[Outputs]
  interval = 1000
  exodus = true
[]

**Comment by jiangwen84** [comment, marked as answer]
You probably want to apply the NeumannBC on variable `c` instead of `w`.

**Reply by souzanha** [reply]
Yes, I've been thinking about that actually. Looking at the Cahn-Hilliard residual, we get a surface term (boundary term) for both the chemical potential and the concentration. The chemical potential and the concentration are both coupled in a way. 

I started out with applying the NeumannBC on c, and did not get a change in the system. Then I found a few examples where other people have applied the flux to w instead. 

In the end, I would like to use the Butler-Volmer BC I have which induces a flux of ions, and looking at the published work I try to mimic the flux has the same units as the chemical potential and not the concentration term. 

I would be very happy to discuss this :) Thanks for your answer.

**Reply by jiangwen84** [reply]
Either way you should see a mass change. I am curious how do you know there is no change in the system. Did you compute the total mass and look at its change?

**Reply by souzanha** [reply]
I'm sure there are better way to do this .. but I display the system in Vislt and look at the last time-step. The color bar stays at max/min=1. Could it be that the change is so small that it isn't displayed? I tried plotting it across a line, and still get c=1.

**Reply by souzanha** [reply]
So I display the variable c as a 2D graph.

**Reply by jiangwen84** [reply]
It is better to use a post processor to compute the integral of c over whole domain. 

I made some tests a while ago. https://github.com/jiangwen84/moose/commit/cf524154e65f0f982c93540053cb70305be7ddc4 

take a look at modules/phase_field/test/tests/CahnHilliardFluxBC/anisotropic_split.i  in that branch.

**Reply by souzanha** [reply]
Thank you, I'll take a look. I saw some posts regarding the previous CahnHilliardFluxBC but couldn't find any files to look at. 

While I have you on the loop, do you mind another question? Its regarding the DerivativeParsedMaterial for the free energy. I have written the function completely underivated (i.e. f_chem(c) and not df/dc that usually goes into the chemical potential), but specified the order to 2. Is this correct?

**Reply by jiangwen84** [reply]
Correct. The code will compute derivatives up the the order you specify.

**Reply by souzanha** [reply]
Thank you for suggesting the postprocessor, its really handy.
This is confusing though, I set the initial condition on c to be 1. However it is 0? It is also increasing (very little) when it actually should decrease due to the negative flux value. 

Postprocessor Values:
+----------------+----------------+
| time           | total_solute   |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
|   1.000000e+00 |   1.000000e-14 |
|   2.000000e+00 |   1.000000e-14 |
|   3.000000e+00 |   1.000000e-14 |
|   4.000000e+00 |   1.000000e-14 |
|   5.000000e+00 |   1.000000e-14 |
|   6.000000e+00 |   1.000000e-14 |
|   7.000000e+00 |   1.000000e-14 |
|   8.000000e+00 |   1.000000e-14 |
|   9.000000e+00 |   1.000000e-14 |
|   1.000000e+01 |   1.000000e-14 |
|   1.100000e+01 |   1.000000e-14 |
|   1.200000e+01 |   1.000000e-14 |
|   1.300000e+01 |   1.000000e-14 |
|   1.400000e+01 |   1.000000e-14 |
+----------------+----------------+

**Reply by jiangwen84** [reply]
From what you showed, the integral of c is 1e-14 and does not change at all. 

Looks like your solution does not actually converge, or not sufficiently. You domain size is too small, and could you change the unit system to make it around 1. It is not necessary but it could make the convergence setup easier. 

You also need to tight the tolerance a bit, i.e.,  
nl_rel_tol = 1.0e-10
nl_abs_tol = 1.0e-10

**Reply by souzanha** [reply]
Now it starts to make sense! So the reason I have such a small system is to compare to the published work. 

Changing the tolerance and the size to 1 -> does not converge. Which is a good thing in a way, now I get that something is wrong while knowing something is wrong.

Do you have any idea for why the initial state doesn't match my input? I'm setting c=1.

Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+
| time           | total_solute   |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+

**Reply by souzanha** [reply]
What do you mean by changing the unit system? Ah maybe by making it all dimensionless, and scaling it? I went with all SI units hoping that it would make more sense while learning more about the model.

**Reply by souzanha** [reply]
Many messages now, but I made all equations dimensionless and scaled by a length unit, energy unit and a timeunit. It does really help with the convergence. However while it does converge, the output still seems strange (however the c is now decreasing which is a good sign).

As a reminder, I set c=1, and a negative flux value = -0.1.


Postprocessor Values:
+----------------+----------------+
| time           | total_solute   |
+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |
|   2.000000e+00 |   9.980000e+03 |
|   4.000000e+00 |   9.960000e+03 |
|   6.000000e+00 |   9.940000e+03 |
|   6.750000e+00 |   9.931375e+03 |
|   8.750000e+00 |   9.910842e+03 |
|   1.075000e+01 |   9.890782e+03 |
|   1.275000e+01 |   9.870776e+03 |
|   1.375000e+01 |   9.861401e+03 |
+----------------+----------------+

**Reply by jiangwen84** [reply]
The postprocessor integrates c over the domain. I guess your domain size is 100 x 100 and at initial, the total value should be 1e4. 

I am not quite sure why at t=0, the postprocessor shows 0. It is very possible that the variable has not been initialized to 1 before the postprossor is called at initial stage.

**Reply by souzanha** [reply]
Yes I figured it out, also included another postprocessor to output the min value so its looking very good now.
Thank you for all the help,  scaling everything and making everything dimensionless will now be a part of my workflow. I wasn't sure it affected the convergence at all, but now I know better.

I also tried the Butler-Volmer BC I created and its working well, there is a flux out and the concentration is decreasing over time. 

A part of the BC is to call for the chemical potential (J=J0(exp(w)-exp(-w)), which I do with:

[BCs]
 [./right]
    type = ButlerVolmerBC
    variable = c
    boundary = right
    w = w #coupled chemical potential
[../]
[]

h-file: 
  const VariableValue & _w;
  const unsigned int _w_var;
C-file:     
    InputParameters params = ADIntegratedBC::validParams();
    params.addRequiredCoupledVar("w", "Coupled chemical potential.");

params.addRequiredCoupledVar("w", "Coupled chemical potential.");
        : ADIntegratedBC(parameters),
        _w(coupledValue("w")),
        _w_var(coupled("w")),
        
I think I haven't seen this type of coupling in a BC before, I've looked at a lot of examples. 
However, if its acting on variable w then the coupling is unnecessary right? But I always get that I need to declare _w[_qp] which is used to define the flux acting on w.

**Reply by xueyang94** [reply]
> Either way you should see a mass change. I am curious how do you know there is no change in the system. Did you compute the total mass and look at its change?

@jiangwen84 I've had a very similar problem with BC. I am using split-KKS kernels. In my case, using NeumannBC or DirichletBC on concentration variables can't change the total concentration, and only using the chemical potential as BC variable can change the total concentration. The NEWTON convergence is good. Do you expect the concentration BC to change the total concentration as well? If so, what do you think might be happening? Thanks.

**Reply by jiangwen84** [reply]
@xueyang94 

Very good question. After some careful thoughts, I think only set nonzero chemical potential flux will change the concentration. The Cahn-Hilliard equation has the conservation form that dc/dt = div J and J = D grad mu. So if J is not zero on the boundary, it will cause mass change. Does this make sense to you?

**Reply by souzanha** [reply]
This is a really good discussion, thank you @jiangwen84 and @xueyang94. 
I have a question regarding how the flux is defined in moose, so by looking at the residual written on the website and writing the weak form of the equation this should be the surface integral governing the flux:

∫  M* c(1-c) ∇* μ*  ⋅n ⋅**dS** [m^2 ]=J[m^2]        (*:unitless)

Or is it actually in differential form, so that

dμ/dn=M* c(1-c) ∇* μ*  ⋅n= J

**Reply by xueyang94** [reply]
> @xueyang94
> 
> Very good question. After some careful thoughts, I think only set nonzero chemical potential flux will change the concentration. The Cahn-Hilliard equation has the conservation form that dc/dt = div J and J = D grad mu. So if J is not zero on the boundary, it will cause mass change. Does this make sense to you?

This totally makes sense to me. As long as mu has a gradient at the boundary, the total concentration will change. Thanks for the explanation!

**Reply by jiangwen84** [reply]
@souzanha The flux is defined in differential form.
